### PR TITLE
dapr cli list components configuration add namespace flag

### LINKED ISSUE #786: dapr list and components cmd should support define "namespace" info
## Expected Behavior
```
# dapr components -k -A
  NAMESPACE    NAME         TYPE             VERSION  SCOPES  CREATED              AGE
  dapr-system  influx       bindings.influx  v1               2021-08-21 11:45.44  2d
  dapr-system  mqtt-pubsub  pubsub.mqtt      v1               2021-08-21 11:46.37  2d
  dapr-system  statestore   state.redis      v1               2021-08-19 11:21.44  4d
  default      statestore   state.redis      v1               2021-08-19 11:07.26  4d
```
```
# dapr list -k --all-namespaces
  NAMESPACE    APP ID     APP PORT  AGE  CREATED
  dapr-system  nodeapp    3000      1d   2021-08-21 15:23.50
  dapr-system  nodeapp    3000      1d   2021-08-21 15:28.08
  dapr-system  nodeapp    3000      3d   2021-08-19 15:07.55
  dapr-system  pythonapp            3d   2021-08-19 15:24.24
```

## Actual Behavior

Not show the namespace info and cant filter with namespace
```
# dapr components -k
  NAME         TYPE             VERSION  SCOPES  CREATED              AGE
  influx       bindings.influx  v1               2021-08-21 11:45.44  2d
  mqtt-pubsub  pubsub.mqtt      v1               2021-08-21 11:46.37  2d
  statestore   state.redis      v1               2021-08-19 11:21.44  4d
  statestore   state.redis      v1               2021-08-19 11:07.26  4d
```
```
# dapr list -k
  APP ID     APP PORT  AGE  CREATED
  nodeapp    3000      1d   2021-08-21 15:23.50
  nodeapp    3000      1d   2021-08-21 15:28.08
  nodeapp    3000      3d   2021-08-19 15:07.55
  pythonapp            3d   2021-08-19 15:24.24
```
## Release Note

RELEASE NOTE:

### PR DESCRIPTION
**Pull request by marviniter**
# Description

dapr list/components/configuration cmd add --namespace and --all-namespace flag for filter define namespace resource,and output list with namespace info.

## Issue reference
closes #786 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [ ] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Review by tanvigour — APPROVED** [review]

**Comment by orizohar** [comment]
@wcs1only @mukundansundar - looks like this has been approved by an approver, is this going into v1.4? asking since we have a docs PR related that we need to know if to merge in

File: pkg/kubernetes/components.go
Comment (artursouza) [inline pkg/kubernetes/components.go:33]: typo

File: cmd/components.go
Comment (artursouza) [inline cmd/components.go:30]: This must be reported as a breaking change since the default behavior is now to list the default namespace or not all.

File: cmd/configurations.go
Comment (artursouza) [inline cmd/configurations.go:30]: Again, a breaking change.

File: cmd/configurations.go
Comment (artursouza) [inline cmd/configurations.go:38]: Update help guide since the first one lists in default namespace.

File: cmd/components.go
Comment (artursouza) [inline cmd/components.go:38]: Mention that this will list only in default namespace,

**Review by artursouza — CHANGES_REQUESTED** [review]
Shouldn't the default behavior remain the same: "all" instead of "default"?

/cc @mukundansundar

File: pkg/kubernetes/components.go
Comment (marviniter) [reply pkg/kubernetes/components.go]: ok，i have fixed it！

File: cmd/components.go
Comment (mukundansundar) [reply cmd/components.go:30]: Following k8s conventions only the default namespace is default :) otherwise a specific all flag needs to be specified. I am ok with this change as long as we clearly mention it as a breaking change and if we are ok with having a breaking change for this.

File: cmd/configurations.go
Comment (mukundansundar) [inline cmd/configurations.go:52]: @artursouza @msfussell are we ok with changing the short hand flags for name and namespace ?? I think this is a breaking change also.

File: cmd/components.go
Comment (mukundansundar) [reply cmd/components.go]: The NAMESPACE property for Dapr will not be affected by this change. This is changing the default behavior of the list, components and configurations command.

**Comment by mukundansundar** [comment]
@marviniter  please resolve the conflicts.

**Comment by msfussell** [comment]
@marviniter - Would really like to get this change in. Can you take a look at the conflicts? Thanks

File: cmd/configurations.go
Comment (msfussell) [reply cmd/configurations.go:62]: Looking at this, there is no shorthand for "name" above. Given this is optional, I am ok with this breaking change for namespace, since this aligns with K8s, but could there be a shorthand for "name"?

**Comment by msfussell** [comment]
@mukundansundar - can you take a look at this PR for v1.6?

File: cmd/configurations.go
Comment (marviniter) [reply cmd/configurations.go:62]: Can use "-N" for "name" for shorthand name?

**Comment by msfussell** [comment]
@mukundansundar @artursouza - Can you look at this PR for v1.6

**Review by mukundansundar — APPROVED** [review]

**Comment by mukundansundar** [comment]
@msfussell @artursouza Are we ok with the breaking changes in the CLI ?

File: cmd/configurations.go
Comment (mukundansundar) [reply cmd/configurations.go:62]: @msfussell, @artursouza Here based on the code, `-n` is changed from `name` to `namespace` which is a breaking change.

Comment (mukundansundar) [reply cmd/configurations.go:62]: cc @yaron2

File: pkg/kubernetes/configurations.go
Comment (mukundansundar) [inline pkg/kubernetes/configurations.go:54]: same change as to listing components.

**Review by mukundansundar — COMMENTED** [review]
@msfussell @artursouza @yaron2 The question here is are we ok with the small breaking changes that we will be introducing to CLI with this change? 
`-n` changed to `namespace` from `name` in one command. 
components and config listed only in a particular namespace unless `-A,--all-namespaces` flag specified whereas the default behaviour now is to list from all namespaces.

**Comment by mukundansundar** [comment]
@yaron2 Can you look at this PR?

**Comment by mukundansundar** [comment]
@yaron2 @artursouza Please review this PR.

**Comment by msfussell** [comment]
> `-n` changed to `namespace` from `name` in one command.

Thinking on this can we use -ns for namespace or does this have to be a single character value? Alternatively I would still prefer to use another character for example -s for namespace rather than break. As the -A I am fine with that change.

**Comment by msfussell** [comment]
@artursouza @yaron2 - Please comment

File: cmd/configurations.go
Comment (artursouza) [reply cmd/configurations.go:62]: I am OK with this breaking change.

**Review by artursouza — DISMISSED** [review]

File: cmd/configurations.go
Comment (artursouza) [reply cmd/configurations.go:62]: In order to respect the deprecation policy, we should (in this PR), keep 100% backwards compatibility and give a warning that "-n" will mean namespace instead of name in the following release.

File: cmd/configurations.go
Comment (mukundansundar) [reply cmd/configurations.go:49]: One option that I want to suggest is to follow Dapr deprecation policy and breaking changes policy. 
In that we put a warning notice for breaking changes or removal 2 releases prior to removing it. 
I suggest for the namespace part, lets leave the default behavior as it existed (all namespaces are queried). Add in the flag -A for all namespaces also. 
print warning log in each command which has this behavior to start using -A for all namespaces and that default behavior will change to querying only the `default` namespace. 

Then after 2 releases, the behavior can be changed. I think this allows users to know that a change will be there. 

cc @yaron2 @artursouza

**Comment by DeepanshuA** [comment]
I will try to summarize the discussion/key-takeaways so far, so that we can try to close open points and aim for including this PR in v1.7:

Currently, in k8s:
1. When "no flag is passed" - only "default" pods related to default namespace are shown
2. When -A flag is passed, "Pods related to All Namespaces" are shown.
3. -n corresponds to listing/ filter / showing only pods related to Namespace passed with this flag.

Currently, in Dapr:
1. No flag: Currently, all namespaces' components are shown
2. -A: Not a valid flag
3. -n: correspponds to name as a filter

What we should strive to achieve:
1. No flag: We can keep showing components related to "all namespaces" and put a warning. From next release(or after 2 releases), behavior can be changed to query only "default" namespace.
2. -A: Start Querying for "all namespaces" with this PR itself.
3. -n: We can keep treating it as "name" filter only and put a warning. From next release(or after 2 releases), behavior can be changed to treat it as "namespace".
4. Do we need a shorthand for "name" as well? I think -name shorthand itself would be good. This can be started from following release. Please confirm.
@artursouza @msfussell @mukundansundar @yaron2  Please confirm about these 4 points.

Warning message which can be printed with every kubernetes mode command to let user know pretty well about the change(or can it be annoying?):
 "From upcoming release: by default to query only "default" namespace and -n shorthand to represent namespace."
 
@marviniter, thanks for this PR. Would it be possible for you to get the chnages incorporated before 1.7 code freeze? Or else, I can attempt making required changes, with your fork as base branch.

**Comment by mukundansundar** [comment]
> 1. No flag: We can keep showing components related to "all namespaces" and put a warning. From next release(or after 2 releases), behavior can be changed to query only "default" namespace.
> 2. -A: Start Querying for "all namespaces" with this PR itself.
> 3. -n: We can keep treating it as "name" filter only and put a warning. From next release(or after 2 releases), behavior can be changed to treat it as "namespace".
> 4. Do we need a shorthand for "name" as well? I think -name shorthand itself would be good. This can be started from following release. Please confirm.

For shorthand for name we can consider `-N`. 
All the other points look good IMO. @artursouza @yaron2 @msfussell any thoughts on the same? 

Warning message which can be printed with every kubernetes mode command to let user know pretty well about the change: 

`In future releases, this command will only query the "default" namespace. Please use the -n (namespace) flag, for specific namespace, or -A (all-namespaces) flag for all namespaces.`

**Comment by DeepanshuA** [comment]
> Warning message which can be printed with every kubernetes mode command to let user know pretty well about the change:
> 
> `In future releases, this command will only query the "default" namespace. Please use the -n (namespace) flag, for specific namespace, or -A (all-namespaces) flag for all namespaces.`

Edited a bit:
 `In future releases, this command will only query the "default" namespace by default. Please use the -n (namespace) flag, for specific namespace, or -A (all-namespaces) flag for all namespaces.`

**Comment by marviniter** [comment]
> I will try to summarize the discussion/key-takeaways so far, so that we can try to close open points and aim for including this PR in v1.7:
> 
> Currently, in k8s:
> 
> 1. When "no flag is passed" - only "default" pods related to default namespace are shown
> 2. When -A flag is passed, "Pods related to All Namespaces" are shown.
> 3. -n corresponds to listing/ filter / showing only pods related to Namespace passed with this flag.
> 
> Currently, in Dapr:
> 
> 1. No flag: Currently, all namespaces' components are shown
> 2. -A: Not a valid flag
> 3. -n: correspponds to name as a filter
> 
> What we should strive to achieve:
> 
> 1. No flag: We can keep showing components related to "all namespaces" and put a warning. From next release(or after 2 releases), behavior can be changed to query only "default" namespace.
> 2. -A: Start Querying for "all namespaces" with this PR itself.
> 3. -n: We can keep treating it as "name" filter only and put a warning. From next release(or after 2 releases), behavior can be changed to treat it as "namespace".
> 4. Do we need a shorthand for "name" as well? I think -name shorthand itself would be good. This can be started from following release. Please confirm.
>    @artursouza @msfussell @mukundansundar @yaron2  Please confirm about these 4 points.
> 
> Warning message which can be printed with every kubernetes mode command to let user know pretty well about the change(or can it be annoying?): "From upcoming release: by default to query only "default" namespace and -n shorthand to represent namespace."
> 
> @marviniter, thanks for this PR. Would it be possible for you to get the changes incorporated before 1.7 code freeze (if any, as per above 4 points finalizations)? Or else, I can attempt making required changes, with your fork/PR code as base.

ok,i'll change it

**Comment by DeepanshuA** [comment]
> ok,i'll change it

thanks @marviniter, highly appreciate you getting back on this

**Comment by mukundansundar** [comment]
@marviniter The code freeze for 1.7 milestone is early next week 22nd. Would you be able to update this PR by then for getting it into 1.7 release?

**Comment by marviniter** [comment]
Hi @mukundansundar @DeepanshuA i had update the PR:
1. No flag: Keep showing components related to "all namespaces" and put a warning
2. -A: Start Querying for "all namespaces"
3. Shorthand for name set to -N
4. Print Warning message when kubernetes mode command

**Comment by mukundansundar** [comment]
@marviniter Since the output for components, configurations and list are changed, can you updated the E2E test for the same?

**Comment by marviniter** [comment]
> @marviniter Since the output for components, configurations and list are changed, can you updated the E2E test for the same?

ok,I am solving this problem

**Comment by DeepanshuA** [comment]
> Hi @mukundansundar @DeepanshuA i had update the PR:
> 
> 1. No flag: Keep showing components related to "all namespaces" and put a warning
> 2. -A: Start Querying for "all namespaces"
> 3. Shorthand for name set to -N
> 4. Print Warning message when kubernetes mode command

@marviniter  As compared to agreed 4 points, I think these points differ in terms of bringing breaking change to -n behaviour from this release itself.  -n will start corresponding to namespace from this release itself, instead of name - without giving first a warning in this release. If that is fine with maintainers, then alright. 
cc @mukundansundar @artursouza

**Comment by mukundansundar** [comment]
> > Hi @mukundansundar @DeepanshuA i had update the PR:
> > 
> > 1. No flag: Keep showing components related to "all namespaces" and put a warning
> > 2. -A: Start Querying for "all namespaces"
> > 3. Shorthand for name set to -N
> > 4. Print Warning message when kubernetes mode command
> 
> @marviniter As compared to agreed 4 points, I think these points differ in terms of bringing breaking change to -n behaviour from this release itself. -n will start corresponding to namespace from this release itself, instead of name - without giving first a warning in this release. If that is fine with maintainers, then alright. cc @mukundansundar @artursouza

@marviniter One other change here is, if `kubernetesMode` flag is set, a warning is printed ... 

```
=== RUN   TestKubernetesNonHAModeMTLSDisabled/install_1.6.0
[193](https://github.com/dapr/cli/runs/5583643277?check_suite_focus=true#step:10:193)
    common.go:468: install without mtls
[194](https://github.com/dapr/cli/runs/5583643277?check_suite_focus=true#step:10:194)
    common.go:474: {"time":"2022-03-17T09:46:19.452215962Z","status":"pending","msg":"Making the jump to hyperspace..."}
[195](https://github.com/dapr/cli/runs/5583643277?check_suite_focus=true#step:10:195)
        {"time":"2022-03-17T09:46:19.452278863Z","status":"info","msg":"Note: To install Dapr using Helm, see here: https://docs.dapr.io/getting-started/install-dapr-kubernetes/#install-with-helm-advanced\n"}
[196](https://github.com/dapr/cli/runs/5583643277?check_suite_focus=true#step:10:196)
        {"time":"2022-03-17T09:46:19.452286463Z","status":"pending","msg":"Deploying the Dapr control plane to your cluster..."}
[197](https://github.com/dapr/cli/runs/5583643277?check_suite_focus=true#step:10:197)
        {"time":"2022-03-17T09:46:39.262188062Z","status":"success","msg":"Deploying the Dapr control plane to your cluster..."}
[198](https://github.com/dapr/cli/runs/5583643277?check_suite_focus=true#step:10:198)
        {"time":"2022-03-17T09:46:39.262264463Z","status":"success","msg":"Success! Dapr has been installed to namespace dapr-cli-tests. To verify, run `dapr status -k' in your terminal. To get started, go here: https://aka.ms/dapr-getting-started"}
[199](https://github.com/dapr/cli/runs/5583643277?check_suite_focus=true#step:10:199)
        {"time":"2022-03-17T09:46:39.262300663Z","status":"warning","msg":"In future releases, this command will only query the \"default\" namespace by default. Please use the -n (namespace) flag, for specific namespace, or -A (all-namespaces) flag for all namespaces."}
```

The above log is for `dapr init -k`. 
This should not be the case. It should only be for the specific commands. 

Also the `-N` flag change for `name` is breaking ... Would like to see if that can be changed back to `-n` with a warning message printed that `-n` will refer to namespace in future. 

cc @artursouza

**Comment by marviniter** [comment]
Hi @mukundansundar @DeepanshuA i had adjust PR as suggested：

No flag: Keep showing components related to "all namespaces" and put a warning
-A: Start Querying for "all namespaces"
Shorthand for name: keep treating it as "name" filter only and put a warning
k8sMode Warning message: only for configurations and compoents commands

File: cmd/components.go
Comment (mukundansundar) [inline cmd/components.go:37]: nit : `-n` is no longer the namespace shorthand.

File: cmd/configurations.go
Comment (mukundansundar) [inline cmd/configurations.go:37]: same as above

File: cmd/list.go
Comment (mukundansundar) [inline cmd/list.go:81]: same as above. change only to .... Please use the --namespace flag, ....

File: cmd/configurations.go
Comment (mukundansundar) [inline cmd/configurations.go:65]: `-n` is no longer namespace shorthand.

File: cmd/components.go
Comment (mukundansundar) [inline cmd/components.go:65]: `-n` is no longer namespace shorthand.

File: tests/e2e/common/common.go
Comment (mukundansundar) [inline tests/e2e/common/common.go:575]:
```suggestion
func componentsOutputCheck(t *testing.T, fields []string, namespace string) {
```

**Review by mukundansundar — COMMENTED** [review]
lgtm. minor nits. Can be fixed after PR is merged also. 
cc @artursouza

**Comment by yaron2** [comment]
We don't have to imitate `kubectl` in regards to param names.

In order to not break the `-n` param usage, we can categorically refer to namespace with `-ns`.

**Comment by mukundansundar** [comment]
> We don't have to imitate `kubectl` in regards to param names.
> 
> In order to not break the `-n` param usage, we can categorically refer to namespace with `-ns`.

I tried initialising this cobra fails if shorthand has more than one char.

File: cmd/configurations.go
Comment (marviniter) [reply cmd/configurations.go:65]: This example use "-n target" for filter name, "target" is resource name,should i change it?

**Review by mukundansundar — APPROVED** [review]
lgtm

**Comment by mukundansundar** [comment]
merging this now ... will be creating a separate PR for fixing the changes mentioned above.

File: cmd/components.go
Comment (mukundansundar) [reply cmd/components.go:37]: fixed in https://github.com/dapr/cli/pull/927

File: cmd/configurations.go
Comment (mukundansundar) [reply cmd/configurations.go:37]: fixed in https://github.com/dapr/cli/pull/927

File: cmd/list.go
Comment (mukundansundar) [reply cmd/list.go:81]: fixed in https://github.com/dapr/cli/pull/927

File: tests/e2e/common/common.go
Comment (mukundansundar) [reply tests/e2e/common/common.go:575]: fixed in https://github.com/dapr/cli/pull/927

**Comment by mukundansundar** [comment]
@marviniter Thanks a lot for this contribution !!

**Comment by DeepanshuA** [comment]
@marviniter thanks for the contribution.
@mukundansundar thanks for reviewing it on priority.

I think we may also need to take care of https://github.com/dapr/docs/pull/1749 to mention new -A flag and may be a warning in general that what changes will be coming in future release.

**Comment by msfussell** [comment]
Please update the docs issue https://github.com/dapr/docs/pull/1749

**Pull request merged**
